### ISSUE TITLE
Electronics Schema

### ISSUE THREAD

**Issue by greghesp** [willblev/UDS_ATC#4]
Do you have a wiring diagram/ scheme for the electronics part?
The software side is simple for me to understand, but the hardware side I'm not so sure on

**Comment by willblev** [comment]
Hi @greghesp, 

I originally intended to upload a diagram/schematic after I got everything working smoothly, but I never got around to it! I doubt that I'll be able to upload a schematic the repo for a while, as I'm not in the same location as Smokey... but if you're interested in making your own, I'd be happy to help figure everything out.

From the code, we can start to see where the different components are connected to the WeMos D1 mini:

```
int fanControlpin = D0;
int const thermoSOpin = D6; // pin D6  *MISO*
int const thermoSCKpin = D5;  // pin D5 *CLK*
int const probe_A_CSpin = D1;  // 
int const probe_B_CSpin = D2; // 
int const probe_C_CSpin = D3; //  
int const probe_D_CSpin =D4; // 
int const probe_E_CSpin = D7; //  
```

As far as I can remember (I'm not able to open it and snap photos as previously mentioned), the thermocouples are all sharing a common connection to D6 (MISO) and D5 (CLK), then each of the thermocouples' CS pins are connected to a different pin on the WeMos. I think that the fanControlpin is wired directly to the PWM input lead of the fan, as the fan model I used could accept a range of voltages on this input so I didn't have to do anything fancy. 

Do you already have the components?

**Comment by greghesp** [comment]
I've got them similar components on order, just waiting for them to arrive.

I've got a NodeMCU rather than a D1 mini, although that should make little difference, and I've ordered some MAX31865 amplifiers. Very similar setup from what I can tell though

**Comment by willblev** [comment]
As long as it's a ESP8266 based board, you should be set. Just be aware
that the pin names may vary between the D1 mini and the board you use, so
you may need to make adjustments to the code to ensure the CLK, MOSI, etc.
pins are going to the correct pins on your board.

**Comment by greghesp** [comment]
Im running a NodeMCU, but U cant even get the MAX31865 example code to upload.  Seems when the max's are connected, it fails

**Comment by willblev** [comment]
Hi again!

So if I have understood correctly, you are using this model of ESP8266-based microcontroller:
![image](https://user-images.githubusercontent.com/5909997/104092130-42826500-5282-11eb-975a-dc0f907fff98.png)
You can find some more details about the [different functions of each pin in the NodeMCU here ](https://www.make-it.ca/nodemcu-arduino/nodemcu-details-specifications/).

My build/code uses the [WeMos D1 mini](https://escapequotes.net/esp8266-wemos-d1-mini-pins-and-diagram/), which looks like this:
![image](https://user-images.githubusercontent.com/5909997/104092189-9725e000-5282-11eb-9049-a65897fc991e.png)

Both boards should be able to run the code, with minor changes such as the names of the pins in the code. You should note that _the name of the pin you need to use in the Arduino IDE is not always what you may expect, so look this up for your MCU_. You may need to map out on paper the allocation of each pin to each component, as you may have conflicts (e.g. you run out of PWM capable GPIOs to run the blower fan if you use the wrong one of the two possible MOSI pins).

I opened up my temperature controller to try and figure out how I have everything wired, but unfortunately I was quite generous with the hot glue, so I couldn't get any good pictures. What I CAN see, is that the MAX31865 share some common connections (SCK, MISO, MOSI) like is shown [here](https://forum.arduino.cc/index.php?topic=461755.0): 
![image](https://user-images.githubusercontent.com/5909997/104092432-3b5c5680-5284-11eb-92a1-c53bd42bf3da.png)


I would recommend that you first try connecting just one thermocouple and get the example code running with that, then try scaling up piece by piece. Hope this helps you to move forward!

**Comment by greghesp** [comment]
Sorted! I had to strip back your code and build it up from scratch selecting what I needed using the MAX31865 library, but I've managed to get the temperatures to read out from both probes now :)

In your `readTempSensors` function, what's the reason for the calculation to set the temperature?

` foodTemperature_Last = (2 * foodTemperature_Last + foodTemperature)/3;`

Wouldn't it just be ` foodTemperature_Last = foodTemperature`

**Comment by willblev** [comment]
Glad to hear that you got it working! You can change the aforementioned line if you want; I think my intention was to smooth the values to avoid turning off everything prematurely due to one high reading. Sort of like a weighted average of the last few readings.

**Comment by greghesp** [comment]
Out of interest, how are you powering it and doing the 12v to 3v?  I've got some buck converters from my 3d printer, but they seem a little big and excessive for this xD

**Comment by willblev** [comment]
I believe I used a cheap (~$1) DC-DC step down voltage regulator, but I
think I stepped down 12v to 5v as my microcontroller could take that as
input. If you check what voltage and current your MAX modules need, you
might only need to use one converter to power the MCU and modules.

**Comment by greghesp** [comment]
Seems the NodeMCU outputs 3v, so have bought a cheap 3v to 12v setup

**Comment by willblev** [comment]
Long story short, I am afraid that taking the 3v output of your NodeMCU and
then converting it up to 12v isn't going to be enough to power your fan :(


You need to consider the load demands such as voltage (Volts) and current
(Amps) as well as the duty cycle needs of your components. A decently sized
blower fan @ 12v will need much more than 1A, and you'll be running it for
up to 15h at a time.

The NodeMCU has an onboard voltage regulator (usually an AMS1117) which
typically takes an input voltage of 5-12v and delivers ~1.5A @ 3.3v. If you
want to power a small peripheral module that can run on 3.3v and only uses
a couple hundred mA (1000mA = 1A), this onboard regulator is sufficient.
But if you need to run something bigger, like a decently sized blower fan @
12v, you'll need much more. We can use the formula below to see how much
current we'd get out from converting from 1.5A @ 3.3v to 12v (assuming our
step-up voltage regulator has an efficiency of 80%, which I just made up,
but just to illustrate my point):

P = U * I

Where P is power in Watts, U is voltage in Volts, and I is current in Amps

Our input is 3.3v @1.5A, and our output is 12v.
3.3v * 1.5A = 4.95 Watts
Since we said we'd lose 20% of the power in the step-up conversion, 4.95 *
0.8 = 3.96 Watts
3.96 Watts / 12v = 0.33A or 330mA

Many 12v fans draw 2-4A at their top speed, so we're nowhere close to being
able to power the fan at full blast using the step-up voltage converter. If
you can, I would power your whole project with a hefty ~4A @12v power
supply, then use a step-down DC-DC converter which can give you ≥1A @5v.
The specs of the onboard NodeMCU voltage regulator looks like it can only
take up to 10v, so you would risk frying it if you didn't step down the
voltage first. 1A @5v should be enough to power the NodeMCU, as well as the
MAX modules (which can run on 3-5v and draw under 50mA each if memory
serves). Double check for yourself before accidentally frying one though ;)

Hope this helps!

**Comment by greghesp** [comment]
Makes sense, its a 12v 3a fan.  Ideally I'd like to power this via USB from a power bank (Micro, Mini, USB C) but I cant seem to find anything to handle this

**Comment by willblev** [comment]
Something like this may work.
https://m.aliexpress.com/item/4000289210562.html?spm=a2g0n.detail.0.0.880e62a02ivq1U&gps-id=storeRecommendH5&scm=1007.18500.187585.0&scm_id=1007.18500.187585.0&scm-url=1007.18500.187585.0&pvid=2c2be017-42d8-492e-9a9b-d67f5341abd9&_t=gps-id%3AstoreRecommendH5%2Cscm-url%3A1007.18500.187585.0%2Cpvid%3A2c2be017-42d8-492e-9a9b-d67f5341abd9%2Ctpp_buckets%3A668%230%23131923%2354_668%23888%233325%239_668%232846%238107%231934_668%232717%237567%23912__668%233374%2315176%23928&browser_id=bad6884fab3b46f983212365fecfadad&aff_trace_key=438c73d1c2144e179d1d3d014ec35048-1603092259884-09918-UneMJZVf&aff_platform=msite&m_page_id=u5jafrevm3scasuo176ed0f66281dd40139c21e326&gclid=&_imgsrc_=ae01.alicdn.com%2Fkf%2FH0f965b03ff6b48a79a9c29d899a7c9a8t.jpg_640x640Q90.jpg_.webp

It's a power bank which uses 6x 18650 lipo cells (sold separately) and is
able to deliver plenty of current over a wide range of voltages (including
12v). I have used it to power my laptop @19v. Decent 18650 cells can often
be recovered from old laptop batteries if you can source those easily- just
be careful and watch a few tutorials so you don't blow anything up!
